Hotcorners settings migration

andrunko opened the pull request:
This re-adds support for changing hotcorner targets (not the panel button) size and position as it was before on eos 3.8.

https://phabricator.endlessm.com/T30445

Commit:
overview: Add flag to avoid showing dialog when no windows are visible

https://phabricator.endlessm.com/T30445

Commit:
hotCorner: Rename indicator class

https://phabricator.endlessm.com/T30445

Commit:
hotCorner: Add specialized HotCorner class for hotcorner targets

https://phabricator.endlessm.com/T30445

Commit:
hotCorner: Add indicator menu to enable/disable hotcorner targets

Note that this setting controls whether hotcorner targets that are
triggerred by holding the pointer on one of the four corners of a screen
(depending on the settings) are enabled but it does not allow disabling
the action that is triggered by clicking on the hotcorner button.

Original code imported from Endless 3.8 gnome-shell.

https://phabricator.endlessm.com/T30445

Commit:
layout: Use specialized HotCorner class for hotcorner targets

Original code imported from Endless 3.8 gnome-shell.

https://phabricator.endlessm.com/T30445

Commit:
Migrate settings

This migrate the old panel settings from Endless 3.8 gnome-shell.

Note that it does not include migrating the 'activate-icon-N'
keybindings as we will now rely on the upstream
'switch-to-application-N' keybindings instead, which use the same
default values.

https://phabricator.endlessm.com/T30445

andrunko commented:
One detail worth mentioning is that on the schema for `hot-corner-enabled` on eos 3.8 the default value was set to false but the description said it was enabled by default - I kept the default value and updated the description instead.

GeorgesStavracas commented on utils.js:
```diff
@@ -22,6 +22,20 @@ const { Gio } = imports.gi;
+
+    return new Gio.Settings({ settings_schema: schema });
+}
+
```
I don't think this is needed. You can simply call `ExtensionUtils.getSettings()`

andrunko replied:
Ok, will try.

andrunko replied:
Done.

GeorgesStavracas commented on ui/overview.js:
```diff
@@ -120,7 +120,7 @@ function _showWindows(overview) {
 }
 
-function toggleApps(overview) {
+function toggleApps(overview, ignoreNoWindowsDialog) {
```
Probably better to pass a default value here: `ignoreNoWindowsDialog = false`

andrunko replied:
Ok, will do - though if we don't pass anything the variable will be "undefined" which serves the same purpose.

andrunko replied:
Done.

andrunko replied:
Sorry, there were actually 2 places that needed updating, now it is really done.

GeorgesStavracas commented on ui/overview.js:
```diff
@@ -130,7 +130,7 @@ function toggleApps(overview) {
     }
 
-    if (!WorkspaceMonitor.hasActiveWindows()) {
+    if (!ignoreNoWindowsDialog && !WorkspaceMonitor.hasActiveWindows()) {
```
You may want to update your WorkspaceMonitor implementation with this: https://github.com/endlessm/eos-desktop-extension/blob/master/ui/workspaceMonitor.js#L87

andrunko replied:
Not sure I get what you mean, care to ellaborate?

GeorgesStavracas replied:
The implementation of WorkspaceMonitor in this extension still uses `Shell.AppSystem`, which is unreliable. The URL I sent shows the more recent and reliable implementation, which inspects the window group tree.

andrunko replied:
I see, I changed (still local) the panel impl to use the window group to detect whether there are visible windows but it still uses AppSystem to check if there are active apps (AppSystem.get_running() > 0) which is what that code path uses. Do you think we should check for visible windows here? I guess not, just want to verify.

GeorgesStavracas replied:
It makes sense to use the same check in both cases. If an app has no window, GNOME Shell won't be tracking it anyway.

andrunko replied:
You mean to use getActiveApps() or getVisibleApps()? I think we want getActiveApps() (as it is now) here because we want to show the window switcher even if all apps are minimized and hasVisibleApps() will return false if all apps are minimized...

andrunko replied:
Updated the workspaceMonitor impl but still using getActiveApps() in this codepath, please check.

GeorgesStavracas commented on schemas/com.endlessm.panel-extension.gschema.xml:
```diff
@@ -1,5 +1,49 @@
+        Default position of the hot corner is in a bottom corner.
+        Set false to move the hot corner to a top corner.
+      </description>
+    </key>
```
This might be a good chance to get rid of this messy composition. It's probably better to introduce an uint `hot-corner-position` key, which would act as a flag (top/down being the first bit, left/right being the second bit)

andrunko replied:
Yes, that could simplify the settings but not sure if it would make it harder for users - I would prefer to not change this now tbh but let me know.

GeorgesStavracas replied:
Which users? The people who are able to reach these GSettings are very likely also able to read the description of the setting (which would list the 4 values here). I don't think this is a reason to introduce convoluted code, or an argument whatsoever.

The reason I think it is a good idea to do it now is that migration didn't happen yet; once this lands, and is released, there's no turning back.

andrunko replied:
Ok, maybe users is not a good argument, but the other argument is that I don't see how having a hot-corner-position instead of 2 keys that are more clear would help making the code less convulated (note the code wouldn't change much, it would basically change the keys that are read and add some complexity to the migration).

andrunko replied:
Lets keep these settings separate for now, if by any chance we want to change this later on we can introduce migration version as you did in the desktop extension.

andrunko replied:
Not changed as explained above.

GeorgesStavracas commented on schemas/com.endlessm.panel-extension.gschema.xml:
```diff
@@ -1,5 +1,49 @@
 <?xml version="1.0" encoding="UTF-8" ?>
 <schemalist>
   <schema path="/com/endlessm/panel-extension/" id="com.endlessm.panel-extension" >
+    <key name="hot-corner-enabled" type="b">
```
There's a similar key under `org.gnome.desktop.interface` named `enable-hot-corners`, it's probably better to either use it.

andrunko replied:
Yes, thought of that but ended up going with the downstream key as this way we can still control whether hotcorner *target* (not the panel button which always work) is enabled without affecting sessions that don't enable the panel extension - but I can change if you prefer.

andrunko replied:
Fwiw, on 3.8 we check both keys and only enable hotcorner if both are true but it has an issue where the hotcorner button menu to enable/disable hotcorners only changes the downstream key. On this PR I changed to only ever check the downstream key and ignore the upstream key.

GeorgesStavracas replied:
I think we discussed this before, and the outcome was: show the button if hot corner is on bottom-right, hide it otherwise.

andrunko replied:
hmm, I think changing the behaviour here to hide the button if not on bottom-right would need some discussions involving the design team, as clicking the button works regardless of whether hotcorner targets are enabled or not (this on 3,8 and the PR here).

I would prefer not to change this behaviour here at all, if you think we should do it please open a separate task for that.

I thought of adding (not included in this PR) a new setting to allow advanced users to hide the button manually, but not doing it automatically...

andrunko replied:
That said I think the best we can do here is to migrate hot-corner-enabled and set the value to the upstream key and use that to determine whether the hotcorner targets are enabled - but even that I am not so sure given the reasons I mentioned above regarding sessions not enabling the panel extension - but still willing to change this one.

andrunko replied:
Updated.

GeorgesStavracas reviewed (changes requested):
Left some comments. I think we can remove some code from here.

GeorgesStavracas commented:
These were my thoughts. Feel free to apply or not them, and merge this

Commit:
workspaceMonitor: Use the window group to determine visible apps

This is more reliable than using the AppSystem.

https://phabricator.endlessm.com/T30445
